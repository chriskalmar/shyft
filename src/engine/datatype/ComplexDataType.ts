export class ComplexDataType {}

export const isComplexDataType = (obj: unknown): obj is ComplexDataType => {
  return obj instanceof ComplexDataType;
};
