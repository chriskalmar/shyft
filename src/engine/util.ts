import * as casual from 'casual';
import * as _ from 'lodash';
import { viewAttributePropertiesWhitelist } from './constants';

type StringFunction = () => string;

export const deleteUndefinedProps = (obj) =>
  Object.keys(obj).forEach((key) => obj[key] === undefined && delete obj[key]);

export const passOrThrow = (
  condition: any,
  messageFn: string | StringFunction,
): void => {
  // providing the error message as a callback return massively improves code coverage reports.
  // the message function is only evaluated if the condition fails, which will show up correctly
  // in the coverage reports.
  if (typeof messageFn !== 'function') {
    throw new Error('passOrThrow() expects messageFn to be a function');
  }
  if (!condition) {
    throw new Error(messageFn());
  }
};

export const resolveFunctionMap = (functionOrMap) => {
  return typeof functionOrMap === 'function'
    ? functionOrMap()
    : { ...functionOrMap };
};

export const isMap = (map: object, nonEmpty?: boolean): boolean => {
  return (
    map !== null &&
    typeof map === 'object' &&
    Array.isArray(map) === false &&
    (!nonEmpty || (nonEmpty && Object.keys(map).length > 0))
  );
};

export const isFunction = (fn: any): boolean => {
  return typeof fn === 'function';
};

export const isArray = (set: any[], nonEmpty?: boolean): boolean => {
  return (
    set !== null &&
    Array.isArray(set) === true &&
    (!nonEmpty || (nonEmpty && set.length > 0))
  );
};

export const isString = (str: any): boolean => typeof str === 'string';

export const mergeMaps = (first, second) => {
  if (!isMap(first) || !isMap(second)) {
    throw new Error('mergeMaps() expects 2 maps for a merge to work');
  }

  return { ...first, ...second };
};

export const mapOverProperties = (
  object: object,
  iteratee: (val: any, key: any) => any,
) => {
  passOrThrow(isMap(object), () => 'Provided object is not a map');

  passOrThrow(
    isFunction(iteratee),
    () => 'Provided iteratee is not a function',
  );

  const keys = Object.keys(object);

  keys.forEach((key) => {
    iteratee(object[key], key);
  });
};

export const sortDataByKeys = (
  keys: string[],
  data: object[],
  keyProperty = 'id',
) => {
  const map = {};
  const result = [];
  const order = {};

  if (!keys || (isArray(keys) && keys.length === 0)) {
    return [];
  }

  if (!data || (isArray(data) && data.length === 0)) {
    return keys.map(() => null);
  }

  for (let d = 0; d < data.length; d++) {
    const row = data[d];
    const id = row[keyProperty];
    const found = map[id];

    if (!found) {
      map[id] = row;
    } else if (found instanceof Array) {
      found.push(row);
    } else {
      map[id] = [found, row];
    }
  }

  for (let k = 0; k < keys.length; k++) {
    const key = keys[k];
    const found = map[key];

    if (typeof found === 'undefined') {
      result.push(null);
    } else if (found instanceof Array) {
      let idx = order[key];

      if (typeof idx === 'undefined') {
        idx = 0;
      } else {
        idx++;
      }

      order[key] = idx;

      result.push(found[idx]);
    } else {
      result.push(found);
    }
  }

  return result;
};

export const reverseString = (str) => {
  return str.split('').reverse().join('');
};

export const randomJson = () => {
  const ret = {};
  const count = casual.integer(1, 5);

  _.times(count, () => {
    const key = _.camelCase(casual.words(2));
    const value = Math.random() > 0.5 ? casual.title : casual.integer();

    ret[key] = value;
  });

  return ret;
};

export const asyncForEach = async (
  array: any[],
  callback: (val, idx, arr) => any,
) => {
  for (let index = 0; index < array.length; index++) {
    await callback(array[index], index, array);
  }
};

export const isDefined = (val: any): boolean => typeof val !== 'undefined';

export const convertEntityToViewAttribute = (attribute) => {
  if (!_.isObject(attribute)) {
    throw new Error(
      'convertEntityToViewAttribute() expects an attribute as input',
    );
  }

  const newAttribute = {};

  viewAttributePropertiesWhitelist.map((prop) => {
    newAttribute[prop] = attribute[prop];
  });

  return newAttribute;
};

export const convertEntityToViewAttributesMap = (attributesMap) => {
  const newAttributesMap = {};

  if (!_.isObject(attributesMap)) {
    throw new Error(
      'convertEntityToViewAttributesMap() expects an attributes map as input',
    );
  }

  for (const [name, attribute] of Object.entries(attributesMap)) {
    newAttributesMap[name] = convertEntityToViewAttribute(attribute);
  }

  return newAttributesMap;
};
