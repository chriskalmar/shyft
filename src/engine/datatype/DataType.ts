import { Entity, ShadowEntity, ViewEntity } from '../..';
import { passOrThrow, isFunction } from '../util';
import { ComplexDataType } from './ComplexDataType';

export type DataTypeSetup = {
  name?: string;
  description?: string;
  mock?: () => any;
  validate?: () => any;
  enforceRequired?: boolean;
  defaultValue?: () => any;
  enforceIndex?: boolean;
};

export type DataTypeFunction = (params: {
  setup: DataTypeSetup;
  entity?: Entity | ShadowEntity | ViewEntity;
}) => DataType | ComplexDataType;

export class DataType {
  name: string;
  description: string;
  mock?: () => any;
  validator?: (value: any, context: any) => any;
  enforceRequired?: boolean;
  defaultValue?: () => any;
  enforceIndex?: boolean;

  constructor(setup: DataTypeSetup = {} as DataTypeSetup) {
    const {
      name,
      description,
      mock,
      validate,
      enforceRequired,
      defaultValue,
      enforceIndex,
    } = setup;

    passOrThrow(name, () => 'Missing data type name');
    passOrThrow(
      description,
      () => `Missing description for data type '${name}'`,
    );

    passOrThrow(
      isFunction(mock),
      () => `'Missing mock function for data type '${name}'`,
    );

    if (validate) {
      passOrThrow(
        isFunction(validate),
        () => `'Invalid validate function for data type '${name}'`,
      );

      this.validator = validate;
    }

    if (defaultValue) {
      passOrThrow(
        isFunction(defaultValue),
        () => `'Invalid defaultValue function for data type '${name}'`,
      );

      this.defaultValue = defaultValue;
    }

    this.name = name;
    this.description = description;
    this.mock = mock;

    if (enforceRequired) {
      this.enforceRequired = enforceRequired;
    }

    if (enforceIndex) {
      this.enforceIndex = enforceIndex;
    }
  }

  validate = async (value: any, context: any): Promise<void> => {
    if (value && this.validator) {
      await this.validator(value, context);
    }
  };

  toString(): string {
    return this.name;
  }
}

export const isDataType = (obj: unknown): obj is DataType => {
  return obj instanceof DataType;
};
