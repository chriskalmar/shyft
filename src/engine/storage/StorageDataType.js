import { passOrThrow, isFunction, isArray } from '../util';

import constants from '../constants';

export class StorageDataType {
  constructor(setup = {}) {
    const {
      name,
      description,
      nativeDataType,
      isSortable,
      serialize,
      enforceSerialize,
      parse,
      capabilities,
    } = setup;

    passOrThrow(name, () => 'Missing storage data type name');
    passOrThrow(
      description,
      () => `Missing description for storage data type '${name}'`,
    );
    passOrThrow(
      nativeDataType,
      () => `Missing native data type for storage data type '${name}'`,
    );

    passOrThrow(
      isFunction(serialize),
      () => `Storage data type '${name}' has an invalid serialize function`,
    );

    passOrThrow(
      !parse || isFunction(parse),
      () => `Storage data type '${name}' has an invalid parse function`,
    );

    if (capabilities) {
      passOrThrow(
        isArray(capabilities),
        () => `Storage data type '${name}' has an invalid list of capabilities`,
      );

      capabilities.map(capability => {
        passOrThrow(
          constants.storageDataTypeCapabilities[capability],
          () =>
            `Storage data type '${name}' has an unknown capability '${capability}'`,
        );
      });
    }

    this.name = name;
    this.description = description;
    this.nativeDataType = nativeDataType;
    this.isSortable = !!isSortable;
    this.serialize = serialize;
    this.enforceSerialize = !!enforceSerialize;
    this.parse = parse || (value => value);
    this.capabilities = capabilities || [];
  }

  toString() {
    return this.name;
  }
}

export const isStorageDataType = obj => {
  return obj instanceof StorageDataType;
};
