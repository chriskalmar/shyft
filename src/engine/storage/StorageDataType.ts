import { passOrThrow, isFunction, isArray } from '../util';

import { storageDataTypeCapabilities } from '../constants';

export type StorageDataTypeSetup = {
  name?: string;
  description?: string;
  // improve nativeDataType typing
  nativeDataType?: any;
  isSortable?: boolean;
  serialize?: Function;
  enforceSerialize?: boolean;
  parse?: Function;
  capabilities?: string[];
};

export class StorageDataType {
  name: string;
  description: string;
  nativeDataType: any;
  isSortable?: boolean;
  serialize: Function;
  enforceSerialize?: boolean;
  parse?: Function;
  capabilities?: string[];

  constructor(setup: StorageDataTypeSetup = {} as StorageDataTypeSetup) {
    const {
      name,
      description,
      nativeDataType,
      isSortable,
      serialize,
      enforceSerialize,
      parse,
      capabilities,
    } = setup;

    passOrThrow(name, () => 'Missing storage data type name');
    passOrThrow(
      description,
      () => `Missing description for storage data type '${name}'`,
    );
    passOrThrow(
      nativeDataType,
      () => `Missing native data type for storage data type '${name}'`,
    );

    passOrThrow(
      isFunction(serialize),
      () => `Storage data type '${name}' has an invalid serialize function`,
    );

    passOrThrow(
      !parse || isFunction(parse),
      () => `Storage data type '${name}' has an invalid parse function`,
    );

    if (capabilities) {
      passOrThrow(
        isArray(capabilities),
        () => `Storage data type '${name}' has an invalid list of capabilities`,
      );

      capabilities.map((capability) => {
        passOrThrow(
          storageDataTypeCapabilities[capability],
          () =>
            `Storage data type '${name}' has an unknown capability '${capability}'`,
        );
      });
    }

    this.name = name;
    this.description = description;
    this.nativeDataType = nativeDataType;
    this.isSortable = !!isSortable;
    this.serialize = serialize;
    this.enforceSerialize = !!enforceSerialize;
    this.parse = parse || ((value: any) => value);
    this.capabilities = capabilities || [];
  }

  toString() {
    return this.name;
  }
}

export const isStorageDataType = (obj: unknown): obj is StorageDataType => {
  return obj instanceof StorageDataType;
};
