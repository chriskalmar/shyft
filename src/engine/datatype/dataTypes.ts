import { DataType } from './DataType';
import { DataTypeUser } from './DataTypeUser';
import { randomJson } from '../util';
import * as casual from 'casual';

export const DataTypeUserID = new DataTypeUser({
  name: 'DataTypeUserID',
  description: 'Data type representing a reference to a user',
  /* istanbul ignore next */
  mock: () => casual.integer(2 ^ 20, 2 ^ 51).toString(),
});

export const DataTypeID = new DataType({
  name: 'DataTypeID',
  description: 'Data type representing unique IDs',
  /* istanbul ignore next */
  mock: () => casual.integer(2 ^ 20, 2 ^ 51).toString(),
});

export const DataTypeInteger = new DataType({
  name: 'DataTypeInteger',
  description: 'Data type representing integer values',
  /* istanbul ignore next */
  mock: () => casual.integer(-2 ^ 10, 2 ^ 10),
});

export const DataTypeBigInt = new DataType({
  name: 'DataTypeBigInt',
  description: 'Data type representing big integer values',
  /* istanbul ignore next */
  mock: () => casual.integer(2 ^ 20, 2 ^ 51).toString(),
});

export const DataTypeFloat = new DataType({
  name: 'DataTypeFloat',
  description: 'Data type representing float values',
  /* istanbul ignore next */
  mock: () => casual.double(-2 ^ 10, 2 ^ 10),
});

export const DataTypeDouble = new DataType({
  name: 'DataTypeDouble',
  description: 'Data type representing double precision values',
  /* istanbul ignore next */
  mock: () => casual.double(-2 ^ 10, 2 ^ 10),
});

export const DataTypeBoolean = new DataType({
  name: 'DataTypeBoolean',
  description: 'Data type representing boolean values',
  /* istanbul ignore next */
  mock: () => casual.boolean,
  enforceRequired: true,
  defaultValue() {
    return false;
  },
});

export const DataTypeString = new DataType({
  name: 'DataTypeString',
  description: 'Data type representing text values',
  /* istanbul ignore next */
  mock: () => casual.title,
});

export const DataTypeJson = new DataType({
  name: 'DataTypeJson',
  description: 'Data type representing json objects',
  /* istanbul ignore next */
  mock: randomJson,
});

export const DataTypeTimestamp = new DataType({
  name: 'DataTypeTimestamp',
  description: 'Data type representing a timestamp',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeTimestampTz = new DataType({
  name: 'DataTypeTimestampTz',
  description: 'Data type representing a timestamp with time zone information',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeDate = new DataType({
  name: 'DataTypeDate',
  description: 'Data type representing a date',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeTime = new DataType({
  name: 'DataTypeTime',
  description: 'Data type representing time',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeTimeTz = new DataType({
  name: 'DataTypeTimeTz',
  description: 'Data type representing time with time zone information',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeUUID = new DataType({
  name: 'DataTypeUUID',
  description: 'Data type representing a UUID',
  /* istanbul ignore next */
  mock: () => casual.uuid,
});

export const DataTypeI18n = new DataType({
  name: 'DataTypeI18n',
  description: 'Data type representing translation objects',
  /* istanbul ignore next */
  mock: randomJson,
});

export const DataTypeUpload = new DataType({
  name: 'DataTypeUpload',
  description: 'Data type representing a file upload',
  mock: () => null,
});
