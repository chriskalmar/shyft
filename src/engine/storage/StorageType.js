import { passOrThrow, isFunction } from '../util';

import { isDataType } from '../datatype/DataType';
import { isStorageDataType } from './StorageDataType';
import { isStorageConfiguration } from './StorageConfiguration';

export class StorageType {
  constructor(setup = {}) {
    const {
      name,
      description,
      findOne,
      findOneByValues,
      find,
      count,
      mutate,
      checkLookupPermission,
    } = setup;

    passOrThrow(name, () => 'Missing storage type name');
    passOrThrow(
      description,
      () => `Missing description for storage type '${name}'`,
    );

    passOrThrow(
      isFunction(findOne),
      () => `Storage type '${name}' needs to implement findOne()`,
    );

    passOrThrow(
      isFunction(findOneByValues),
      () => `Storage type '${name}' needs to implement findOneByValues()`,
    );

    passOrThrow(
      isFunction(find),
      () => `Storage type '${name}' needs to implement find()`,
    );

    passOrThrow(
      isFunction(count),
      () => `Storage type '${name}' needs to implement count()`,
    );

    passOrThrow(
      isFunction(mutate),
      () => `Storage type '${name}' needs to implement mutate()`,
    );

    passOrThrow(
      isFunction(checkLookupPermission),
      () => `Storage type '${name}' needs to implement checkLookupPermission()`,
    );

    this.name = name;
    this.description = description;
    this.findOne = findOne;
    this.findOneByValues = findOneByValues;
    this.find = find;
    this.count = count;
    this.mutate = mutate;
    this.checkLookupPermission = checkLookupPermission;

    this._dataTypeMap = {};
    this._dynamicDataTypeMap = [];
  }

  addDataTypeMap(schemaDataType, storageDataType) {
    passOrThrow(
      isDataType(schemaDataType),
      () =>
        `Provided schemaDataType is not a valid data type in '${this.name}', ` +
        `got this instead: ${String(schemaDataType)}`,
    );

    passOrThrow(
      isStorageDataType(storageDataType),
      () =>
        `Provided storageDataType for '${String(
          schemaDataType,
        )}' is not a valid storage data type in '${this.name}', ` +
        `got this instead: ${String(storageDataType)}`,
    );

    passOrThrow(
      !this._dataTypeMap[schemaDataType.name],
      () =>
        `Data type mapping for '${
          schemaDataType.name
        }' already registered with storage type '${this.name}'`,
    );

    this._dataTypeMap[schemaDataType.name] = storageDataType;
  }

  addDynamicDataTypeMap(schemaDataTypeDetector, storageDataType) {
    passOrThrow(
      isFunction(schemaDataTypeDetector),
      () =>
        `Provided schemaDataTypeDetector is not a valid function in '${
          this.name
        }', ` + `got this instead: ${String(schemaDataTypeDetector)}`,
    );

    passOrThrow(
      isStorageDataType(storageDataType) || isFunction(storageDataType),
      () =>
        `Provided storageDataType for '${String(
          schemaDataTypeDetector,
        )}' is not a valid storage data type or function in '${this.name}', ` +
        `got this instead: ${String(storageDataType)}`,
    );

    this._dynamicDataTypeMap.push({
      schemaDataTypeDetector,
      storageDataType,
    });
  }

  convertToStorageDataType(schemaDataType) {
    const foundDynamicDataType = this._dynamicDataTypeMap.find(
      ({ schemaDataTypeDetector }) => schemaDataTypeDetector(schemaDataType),
    );
    if (foundDynamicDataType) {
      const storageDataType = foundDynamicDataType.storageDataType;

      if (isFunction(storageDataType)) {
        const attributeType = schemaDataType;
        return storageDataType(attributeType);
      }

      return storageDataType;
    }

    passOrThrow(
      isDataType(schemaDataType),
      () =>
        `Provided schemaDataType is not a valid data type in storage type '${
          this.name
        }', ` + `got this instead: ${String(schemaDataType)}`,
    );

    passOrThrow(
      this._dataTypeMap[schemaDataType.name],
      () =>
        `No data type mapping found for '${
          schemaDataType.name
        }' in storage type '${this.name}'`,
    );

    return this._dataTypeMap[schemaDataType.name];
  }

  setStorageConfiguration(storageConfiguration) {
    passOrThrow(
      isStorageConfiguration(storageConfiguration),
      () => 'StorageType expects a valid storageConfiguration',
    );

    this.storageConfiguration = storageConfiguration;
  }

  getStorageConfiguration() {
    passOrThrow(
      this.storageConfiguration,
      () => 'StorageType is missing a valid storageConfiguration',
    );

    return this.storageConfiguration;
  }

  getStorageInstance() {
    return this.getStorageConfiguration().getStorageInstance();
  }

  getStorageModels() {
    return this.getStorageConfiguration().getStorageModels();
  }

  toString() {
    return this.name;
  }
}

export const isStorageType = obj => {
  return obj instanceof StorageType;
};
