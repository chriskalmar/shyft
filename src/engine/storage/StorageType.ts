import { isFunction, passOrThrow } from '../util';

import { DataType, isDataType } from '../datatype/DataType';
import { isStorageDataType } from './StorageDataType';
import { isStorageConfiguration, StorageConfiguration } from './StorageConfiguration';
import { ComplexDataType, Entity, StorageDataType } from '../..';
import { Mutation } from '../mutation/Mutation';
import { Connection } from 'typeorm';

export type StorageTypeSetup = {
  name?: string;
  description?: string;
  findOne?: (
    entity?: Entity,
    id?: any,
    args?: Record<string, any>,
    context?: Record<string, any>,
  ) => any;
  findOneByValues?: (
    entity?: Entity,
    arg?: any,
    context?: Record<string, any>,
  ) => any;
  find?: (
    entity?: Entity,
    args?: Record<string, any>,
    context?: Record<string, any>,
    parentConnection?: any,
  ) => any;
  count?: (
    entity?: Entity,
    args?: Record<string, any>,
    context?: Record<string, any>,
    parentConnection?: any,
  ) => number | any;
  mutate?: (
    entity?: Entity,
    id?: any,
    input?: any,
    entityMutation?: Mutation,
    context?: Record<string, any>,
  ) => void | any;
  checkLookupPermission?: (
    entity?: Entity,
    where?: any,
    context?: Record<string, any>,
  ) => boolean | any;
};

export class StorageType {
  name: string;
  description: string;
  findOne: (...arg) => any;
  findOneByValues: (...arg) => any;
  find: (...args) => any;
  count: (...arg) => any;
  mutate: (...args) => any;
  checkLookupPermission: (...arg) => any;

  private _dataTypeMap;
  private _dynamicDataTypeMap;
  storageConfiguration: StorageConfiguration;

  constructor(setup: StorageTypeSetup = {} as StorageTypeSetup) {
    const {
      name,
      description,
      findOne,
      findOneByValues,
      find,
      count,
      mutate,
      checkLookupPermission,
    } = setup;

    passOrThrow(name, () => 'Missing storage type name');
    passOrThrow(
      description,
      () => `Missing description for storage type '${name}'`,
    );

    passOrThrow(
      isFunction(findOne),
      () => `Storage type '${name}' needs to implement findOne()`,
    );

    passOrThrow(
      isFunction(findOneByValues),
      () => `Storage type '${name}' needs to implement findOneByValues()`,
    );

    passOrThrow(
      isFunction(find),
      () => `Storage type '${name}' needs to implement find()`,
    );

    passOrThrow(
      isFunction(count),
      () => `Storage type '${name}' needs to implement count()`,
    );

    passOrThrow(
      isFunction(mutate),
      () => `Storage type '${name}' needs to implement mutate()`,
    );

    passOrThrow(
      isFunction(checkLookupPermission),
      () => `Storage type '${name}' needs to implement checkLookupPermission()`,
    );

    this.name = name;
    this.description = description;
    this.findOne = findOne;
    this.findOneByValues = findOneByValues;
    this.find = find;
    this.count = count;
    this.mutate = mutate;
    this.checkLookupPermission = checkLookupPermission;

    this._dataTypeMap = {};
    this._dynamicDataTypeMap = [];
  }

  addDataTypeMap(schemaDataType: DataType, storageDataType: StorageDataType) {
    passOrThrow(
      isDataType(schemaDataType),
      () =>
        `Provided schemaDataType is not a valid data type in '${this.name}', ` +
        `got this instead: ${String(schemaDataType)}`,
    );

    passOrThrow(
      isStorageDataType(storageDataType),
      () =>
        `Provided storageDataType for '${String(
          schemaDataType,
        )}' is not a valid storage data type in '${this.name}', ` +
        `got this instead: ${String(storageDataType)}`,
    );

    passOrThrow(
      !this._dataTypeMap[schemaDataType.name],
      () =>
        `Data type mapping for '${schemaDataType.name}' already registered with storage type '${this.name}'`,
    );

    this._dataTypeMap[schemaDataType.name] = storageDataType;
  }

  addDynamicDataTypeMap(
    schemaDataTypeDetector: (...args) => any,
    storageDataType: StorageDataType | ((...args) => any),
  ): void {
    passOrThrow(
      isFunction(schemaDataTypeDetector),
      () =>
        `Provided schemaDataTypeDetector is not a valid function in '${this.name}', ` +
        `got this instead: ${String(schemaDataTypeDetector)}`,
    );

    passOrThrow(
      isStorageDataType(storageDataType) || isFunction(storageDataType),
      () =>
        `Provided storageDataType for '${String(
          schemaDataTypeDetector,
        )}' is not a valid storage data type or function in '${this.name}', ` +
        `got this instead: ${String(storageDataType)}`,
    );

    this._dynamicDataTypeMap.push({
      schemaDataTypeDetector,
      storageDataType,
    });
  }

  convertToStorageDataType(
    schemaDataType: DataType | ComplexDataType,
  ): StorageDataType {
    const foundDynamicDataType = this._dynamicDataTypeMap.find(
      ({ schemaDataTypeDetector }) => schemaDataTypeDetector(schemaDataType),
    );

    if (foundDynamicDataType) {
      const storageDataType = foundDynamicDataType.storageDataType;

      if (isFunction(storageDataType)) {
        return storageDataType(schemaDataType);
      }

      return storageDataType;
    }

    if (!isDataType(schemaDataType)) {
      throw new Error(
        `Provided schemaDataType is not a valid data type in storage type '${
          this.name
        }', got this instead: ${String(schemaDataType)}`,
      );
    }

    passOrThrow(
      this._dataTypeMap[schemaDataType.name],
      () =>
        `No data type mapping found for '${schemaDataType.name}' in storage type '${this.name}'`,
    );

    return this._dataTypeMap[schemaDataType.name];
  }

  setStorageConfiguration(storageConfiguration: StorageConfiguration): void {
    passOrThrow(
      isStorageConfiguration(storageConfiguration),
      () => 'StorageType expects a valid storageConfiguration',
    );

    this.storageConfiguration = storageConfiguration;
  }

  getStorageConfiguration(): StorageConfiguration {
    passOrThrow(
      this.storageConfiguration,
      () => 'StorageType is missing a valid storageConfiguration',
    );

    return this.storageConfiguration;
  }

  getStorageInstance(): Connection {
    return this.getStorageConfiguration().getStorageInstance();
  }

  getStorageModels() {
    return this.getStorageConfiguration().getStorageModels();
  }

  toString(): string {
    return this.name;
  }
}

export const isStorageType = (obj: unknown): boolean => {
  return obj instanceof StorageType;
};
