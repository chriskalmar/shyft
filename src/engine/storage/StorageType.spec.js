/* eslint-disable @typescript-eslint/no-empty-function */

import { StorageType, isStorageType } from './StorageType';
import { StorageDataType } from './StorageDataType';
import { passOrThrow } from '../util';

import { StorageTypeNull } from './StorageTypeNull';

import { DataTypeID, DataTypeString } from '../datatype/dataTypes';

describe('StorageType', () => {
  it('should have a name', () => {
    function fn() {
      new StorageType(); // eslint-disable-line no-new
    }

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  it('should have a description', () => {
    function fn() {
      // eslint-disable-next-line no-new
      new StorageType({
        name: 'Example',
      });
    }

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  it("should return it's name", () => {
    const SomeStorageType = new StorageType({
      name: 'SomeStorageType',
      description: 'Just some description',
      findOne() {},
      findOneByValues() {},
      find() {},
      count() {},
      mutate() {},
      checkLookupPermission() {},
    });

    expect(SomeStorageType.name).toBe('SomeStorageType');
    expect(String(SomeStorageType)).toBe('SomeStorageType');
  });

  it('should throw on missing data fetching implementations', () => {
    let fn;

    fn = () => {
      // eslint-disable-next-line no-new
      new StorageType({
        name: 'SomeStorageType',
        description: 'Just some description',
      });
    };

    expect(fn).toThrowErrorMatchingSnapshot();

    fn = () => {
      // eslint-disable-next-line no-new
      new StorageType({
        name: 'SomeStorageType',
        description: 'Just some description',
        findOne() {},
      });
    };

    expect(fn).toThrowErrorMatchingSnapshot();

    fn = () => {
      // eslint-disable-next-line no-new
      new StorageType({
        name: 'SomeStorageType',
        description: 'Just some description',
        findOne() {},
        findOneByValues() {},
      });
    };

    expect(fn).toThrowErrorMatchingSnapshot();

    fn = () => {
      // eslint-disable-next-line no-new
      new StorageType({
        name: 'SomeStorageType',
        description: 'Just some description',
        findOne() {},
        findOneByValues() {},
        find() {},
      });
    };

    expect(fn).toThrowErrorMatchingSnapshot();

    fn = () => {
      // eslint-disable-next-line no-new
      new StorageType({
        name: 'SomeStorageType',
        description: 'Just some description',
        findOne() {},
        findOneByValues() {},
        find() {},
        count() {},
      });
    };

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  describe('data type mapping', () => {
    const SomeStorageType = new StorageType({
      name: 'SomeStorageType',
      description: 'Just some description',
      findOne() {},
      findOneByValues() {},
      find() {},
      count() {},
      mutate() {},
      checkLookupPermission() {},
    });

    const StorageDataTypeText = new StorageDataType({
      name: 'StorageDataTypeText',
      description: 'Just some description',
      nativeDataType: 'text',
      serialize() {},
    });

    it('should accept new mappings', () => {
      SomeStorageType.addDataTypeMap(DataTypeID, StorageDataTypeText);
      SomeStorageType.addDataTypeMap(DataTypeString, StorageDataTypeText);
    });

    it('should reject invalid schema data types', () => {
      function fn() {
        SomeStorageType.addDataTypeMap({ some: 'thing' }, StorageDataTypeText);
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });

    it('should reject invalid storage data types', () => {
      function fn() {
        SomeStorageType.addDataTypeMap(DataTypeString, { someThing: 'else' });
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });

    it('should throw on duplicate mappings', () => {
      function fn() {
        SomeStorageType.addDataTypeMap(DataTypeID, StorageDataTypeText);
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });
  });

  describe('isStorageType', () => {
    it('should recognize objects of type StorageType', () => {
      const SomeStorageType = new StorageType({
        name: 'SomeStorageType',
        description: 'Just some description',
        findOne() {},
        findOneByValues() {},
        find() {},
        count() {},
        mutate() {},
        checkLookupPermission() {},
      });

      function fn() {
        passOrThrow(
          isStorageType(SomeStorageType) && isStorageType(StorageTypeNull),
          () => 'This error will never happen',
        );
      }

      expect(fn).not.toThrow();
    });

    it('should recognize non-StorageType objects', () => {
      function fn() {
        passOrThrow(
          isStorageType({}) ||
            isStorageType(function test() {}) ||
            isStorageType(Error),
          () => 'Not an StorageType object',
        );
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });
  });
});
