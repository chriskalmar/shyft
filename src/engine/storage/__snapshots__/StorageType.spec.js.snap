// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StorageType data type mapping should reject invalid schema data types 1`] = `"Provided schemaDataType is not a valid data type in 'SomeStorageType', got this instead: [object Object]"`;

exports[`StorageType data type mapping should reject invalid storage data types 1`] = `"Provided storageDataType for 'DataTypeString' is not a valid storage data type in 'SomeStorageType', got this instead: [object Object]"`;

exports[`StorageType data type mapping should throw on duplicate mappings 1`] = `"Data type mapping for 'DataTypeID' already registered with storage type 'SomeStorageType'"`;

exports[`StorageType isStorageType should recognize non-StorageType objects 1`] = `"Not an StorageType object"`;

exports[`StorageType should have a description 1`] = `"Missing description for storage type 'Example'"`;

exports[`StorageType should have a name 1`] = `"Missing storage type name"`;

exports[`StorageType should throw on missing data fetching implementations 1`] = `"Storage type 'SomeStorageType' needs to implement findOne()"`;

exports[`StorageType should throw on missing data fetching implementations 2`] = `"Storage type 'SomeStorageType' needs to implement findOneByValues()"`;

exports[`StorageType should throw on missing data fetching implementations 3`] = `"Storage type 'SomeStorageType' needs to implement find()"`;

exports[`StorageType should throw on missing data fetching implementations 4`] = `"Storage type 'SomeStorageType' needs to implement count()"`;

exports[`StorageType should throw on missing data fetching implementations 5`] = `"Storage type 'SomeStorageType' needs to implement mutate()"`;
